### PR TITLE
Exists method added

### PR DESCRIPTION
**Pull request by elisheva-qlogic**
exists method added in BigtableTableAdminClient to check if table exists #3636

### CODE REVIEW

File: google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java [hunk 1]
```diff
@@ -336,17 +336,20 @@ public void deleteTable(String tableId) {
...
-      if (name.equals(tableId)) {
-        return true;
+    return awaitFuture(existsAsync(tableId)).booleanValue();
+  }
+
+  public ApiFuture<Boolean> existsAsync(final String tableId) {
+    return ApiFutures.transform(listTablesAsync(), new ApiFunction<List<TableName>, Boolean>() {
+      @Override public Boolean apply(List<TableName> tableNames) {
+        for(TableName tableName : tableNames) {
+          if(tableName.getTable().equalsIgnoreCase(tableId)) {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java]: Why ignore case?

File: google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java [hunk 2]
```diff
@@ -454,7 +457,11 @@ public Table getTable(String tableId) {
   // TODO(igorbernstein2): consider changing this method to use relative table ids.
   @SuppressWarnings("WeakerAccess")
   public ApiFuture<List<TableName>> listTablesAsync() {
-    ListTablesRequest request = ListTablesRequest.newBuilder().setParent(instanceName.toString())
+    return listTablesAsync(com.google.bigtable.admin.v2.Table.View.NAME_ONLY);
+  }
+
+  public ApiFuture<List<TableName>> listTablesAsync(com.google.bigtable.admin.v2.Table.View view) {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java]: The view parameter should not be exposed

File: google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java [hunk 3]
```diff
@@ -324,6 +324,34 @@ public void deleteTable(String tableId) {
...
+   *   "Table exists";
+   * }
+   * }</pre>
+   */
+  public boolean exists(String tableId) {
+    return awaitFuture(existsAsync(tableId)).booleanValue();
+  }
+
+  public ApiFuture<Boolean> existsAsync(final String tableId) {
+    return ApiFutures.transform(listTablesAsync(), new ApiFunction<List<TableName>, Boolean>() {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java]: I apologize, I didn't notice this earlier: the exists method should use GetTable using NAME_ONLY view. So it should be something like:

```java
GetTableRequest request = GetTableRequest.newBuilder().setName(tableName).setView(NAME_ONLY).build();

ApiFuture<Table> protoFuture = stub.getTableCallable().futureCall(request);
ApiFuture<Boolean> existsFuture = ApiFutures.transform(protoFuture, new ApiFunction<Table, Boolean>() {
  Boolean apply(Table ignored) { return true; }
})

return ApiFutures.catching(existsFuture, NotFoundException.class, new ApiFunction<NotFoundException, Boolean>() { Boolean apply(Throwable ignored) { return false});

```

File: google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java [hunk 4]
```diff
@@ -324,6 +325,38 @@ public void deleteTable(String tableId) {
...
+   * if(client.exists("my-table")) {
+   *   "Table exists";
+   * }
+   * }</pre>
+   */
+  public boolean exists(String tableId) {
+    return awaitFuture(existsAsync(tableId)).booleanValue();
+  }
+
+  public ApiFuture<Boolean> existsAsync(String tableId) {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java]: Please add javadoc for the async method

File: google-cloud-clients/google-cloud-bigtable-admin/src/test/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClientTest.java [hunk 5]
```diff
@@ -303,4 +303,34 @@ public void testAwaitReplication() {
...
+
+  @Test
+  public void testExistsTrue() {
+    // Setup
+    com.google.bigtable.admin.v2.Table expectedResponse =
+                com.google.bigtable.admin.v2.Table.newBuilder()
+                    .setName(TABLE_NAME.toString())
+                    .build();
+
+    Mockito.when(mockGetTableCallable.futureCall(Matchers.any(GetTableRequest.class))).thenReturn(ApiFutures.immediateFuture(expectedResponse));
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable-admin/src/test/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClientTest.java]: Please format the code according to google java style:
http://google.github.io/styleguide/javaguide.html

You can find formatters for ides here:
https://github.com/google/styleguide

File: google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java [hunk 6]
```diff
@@ -324,6 +325,61 @@ public void deleteTable(String tableId) {
...
+   * <p>Sample code:
+   *
+   * <pre>{@code
+   * ApiFuture<Boolean> found = client.existsAsync("my-table");
+   *
+   * ApiFutures.addCallback(
+   *  found,
+   *  new ApiFutureCallback<Boolean>() {
+   *    public void onSuccess(Boolean found) {
+   *      System.out.println("Table exists");
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java]:
```java
if (found) {
  System.out.println("Table exists");
} else {
  System.out.println("Table not found");
}
```

File: google-cloud-clients/google-cloud-bigtable-admin/src/test/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClientTest.java [hunk 7]
```diff
@@ -303,4 +310,38 @@ public void testAwaitReplication() {
...
+  }
+
+  @Test
+  public void testExistsFalse() {
+    // Setup
+    NotFoundException exception =
+        new NotFoundException("fake error", null, GrpcStatusCode.of(Status.Code.NOT_FOUND), false);
+
+    Mockito.when(mockGetTableCallable.futureCall(Matchers.any(GetTableRequest.class)))
+        .thenReturn(ApiFutures.immediateFailedFuture(exception));
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable-admin/src/test/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClientTest.java]: @elisheva-qlogic this is failing to build, can you fix it with:
```
.thenReturn(ApiFutures.<com.google.bigtable.admin.v2.Table>immediateFailedFuture(exception));
```

**Comment by sduskis** [comment]
@elisheva-qlogic, there's a compilation failure:

> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.5.1:compile (default-compile) on project google-cloud-bigtable-admin: Compilation failure
[ERROR] /tmpfs/src/github/google-cloud-java/google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java:[339,12] cannot find symbol
[ERROR]   symbol:   method awaitFuture(com.google.api.core.ApiFuture<java.lang.Boolean>)
[ERROR]   location: class com.google.cloud.bigtable.admin.v2.BigtableTableAdminClient

It looks like a merge conflict with PR #3809

**Pull request merged**